support/oidc-discovery-provider/cert_manager_test.go

```diff
@@ -0,0 +1,334 @@
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"os"
+	"reflect"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/sirupsen/logrus/hooks/test"
+	"github.com/spiffe/spire/pkg/common/x509util"
+	"github.com/spiffe/spire/test/testkey"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	oidcServerKey    = testkey.MustEC256()
+	oidcServerKeyNew = testkey.MustEC256()
+	certFilePath     = "/oidcServerCert.pem"
+	keyFilePath      = "/oidcServerKey.pem"
+)
+
+func TestTLSConfig(t *testing.T) {
+	logger, logHook := test.NewNullLogger()
+
+	oidcServerKeyDer, err := x509.MarshalECPrivateKey(oidcServerKey)
+	require.NoError(t, err)
+
+	certTmpl := &x509.Certificate{
+		SerialNumber: big.NewInt(0),
+		NotAfter:     time.Now().Add(time.Hour),
+		Subject: pkix.Name{
+			Country:    []string{"BR"},
+			CommonName: "oidc-provider-discovery.example.com",
+		},
+	}
+	oidcServerCert, err := x509util.CreateCertificate(certTmpl, certTmpl, oidcServerKey.Public(), oidcServerKey)
+	require.NoError(t, err)
+	require.NotNilf(t, oidcServerCert, "oidcServerCert is nil")
+
+	oidcServerKeyPem := pem.EncodeToMemory(&pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: oidcServerKeyDer,
+	})
+
+	oidcServerCertPem := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: oidcServerCert.Raw,
+	})
+
+	certTmpl.Subject.Country = []string{"AR"}
+	oidcServerCertUpdated1, err := x509util.CreateCertificate(certTmpl, certTmpl, oidcServerKey.Public(), oidcServerKey)
+	require.NoError(t, err)
+
+	oidcServerKeyNewDer, err := x509.MarshalECPrivateKey(oidcServerKeyNew)
+	require.NoError(t, err)
+
+	oidcServerCertUpdated2, err := x509util.CreateCertificate(certTmpl, certTmpl, oidcServerKeyNew.Public(), oidcServerKeyNew)
+	require.NoError(t, err)
+
+	certTmpl.Subject.Country = []string{"US"}
+
+	tmpDir := t.TempDir()
+
+	writeFile(t, tmpDir+keyFilePath, oidcServerKeyPem)
+	writeFile(t, tmpDir+certFilePath, oidcServerCertPem)
+	writeFile(t, tmpDir+"/oidcServerKeyInvalid.pem", []byte{1})
+	writeFile(t, tmpDir+"/oidcServerCertInvalid.pem", []byte{1})
+
+	chInfo := &tls.ClientHelloInfo{
+		ServerName: "oidc-provider-discovery.example.com",
+	}
+
+	ctx, cancelFn := context.WithCancel(context.Background())
+	certManager, err := NewDiskCertManager(&ServingCertFileConfig{
+		CertFilePath:     tmpDir + certFilePath,
+		KeyFilePath:      tmpDir + keyFilePath,
+		FileSyncInterval: 10 * time.Millisecond,
```

Comment:
This could make the test run faster, but we would still need to use require.Eventually to make some assertions since the time difference between loading a cert and moving the clock + making an assertion can cause errors.